### ISSUE TITLE
BarChart using barSpace = 0 problem

### ISSUE THREAD

**Issue by gal-orlanczyk** [danielgindi/ios-charts#666]
Hi,

I am trying to create a graph for sleeping data like this: 
![screen322x572](https://cloud.githubusercontent.com/assets/6908583/12264772/9b5684c0-b941-11e5-8763-7bc1d5b028c5.jpeg)

So for this i am using bar stack bar chart with 3 types of colors and only 1 at a time gets a value, the problem is even though i am using barSpace = 0 i am getting little spaces between values when i have more than 10 entries, for example when i have 100 this is what i get:
<img width="295" alt="screen shot 2016-01-12 at 15 34 14" src="https://cloud.githubusercontent.com/assets/6908583/12264822/fc45fcc0-b941-11e5-9ccd-7d29e28068e1.png">

Is there any way to avoid the small separating lines between each bar entry?

Thanks,
Gal

**Comment by liuxuan30** [comment]
The very thin white line is what I can reproduce with ChartsDemo; but the bigger spaces, seems like you don't have data there.
Anyway, if you are talking about the very small spaces, I'm not sure if this is related to the pixel and point is not aligned? What you guys think @danielgindi @petester42

**Comment by gal-orlanczyk** [comment]
Yes i am talking about the small thin line, it looks like the rendering when having to draw a lot of bars is wrong. when there is small amount all is good, from 20+ the situation may defer. 
when i give is bar space minus values for example: -0.2 this helps a little, but the problem is all the spacing gets smaller by 20% and again with large amount of bars 600 for example this also doesn't help.

Thanks for the help,
Gal

**Comment by danielgindi** [comment]
This seems to be caused by the way that Core Graphics work. It is indeed a bug- that I hope to find a way to avoid.

**Comment by liuxuan30** [comment]
Not sure how we can align the pixel perfectly

**Comment by danielgindi** [comment]
Well, I plan to sit one day for a few hours and start sketching up a 2D renderer on the GPU. That will definitely solve this issue. As this is what Android does...

**Comment by kyo313** [comment]
So I need to wait, looking for your update

**Comment by philipengberg** [comment]
@liuxuan30 @danielgindi 

It's an issue with all stacked bar charts it seems. If you look at the stacked bar char demo and zoom in, there is a transparent 1 point gap between each color:

![simulator screen shot - iphone 8 - 2017-10-26 at 16 26 45 copy](https://user-images.githubusercontent.com/1840093/32073072-1e429f9e-ba95-11e7-9274-25dd127033c6.png)

You have to zoom way in, but you can even see the vertical gridline zipping through right in the middle of the bar.

This is acceptable for the average use case. But I'm in a situation where I am simulating gradient transitions between colors, and in that case it ends up looking like this:

![simulator screen shot - iphone 8 - 2017-10-26 at 15 37 08 copy](https://user-images.githubusercontent.com/1840093/32073164-6c0c7704-ba95-11e7-8ff8-73fc397029ea.png)

Not great, obviously. So I dug a bit deeper into the belly of the beast and it does indeed seem to be a rounding error, probably caused in the `rectValuesToPixel` function.

I also sort of fixed it. I realized that the bars are drawn top down, so if we just make the height of each bar a tiny bit taller, the issue is solved

```swift
context.fill(CGRect(x: barRect.origin.x, y: barRect.origin.y, width: barRect.width, height: barRect.height + 1))
```

Instead of adding an entire point, you could also you the pixel/point ratio for the current screen I guess.

I know it's a bit hacky, but I don't currently see another way around this.

This is the end result:

![simulator screen shot - iphone 8 - 2017-10-26 at 19 55 02 copy](https://user-images.githubusercontent.com/1840093/32073317-eb0d5e2e-ba95-11e7-85b7-bb04e4d5cc9a.png)

Much better 👍 

What do you guys think? Is it a fix you would accept in a PR?

**Comment by liuxuan30** [comment]
if it's a bug, we definitely accept a PR. But I am not sure `height: barRect.size.height + 1` will solve it in any scenario?

**Comment by philipengberg** [comment]
I'm not sure either, but I couldn't figure out a better way to be honest 😞 And now I have to have the entire library cloned into the instead of using cocoapods just for this one tiny fix. Do you have any other ideas @liuxuan30?

**Comment by petester42** [comment]
You can point cocoapods to your own fork if that is a problem you need fixed right away. As far as the fix goes what happens with 1x, 2x and 3x screens, the `+1` might not work for all.
